lib/SimpleSAML/Modules/Module.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+
+namespace SimpleSAML\Modules;
+
+
+use Symfony\Component\Config\Exception\FileLocatorFileNotFoundException;
+use Symfony\Component\Config\FileLocator;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Loader\DirectoryLoader;
+use Symfony\Component\HttpKernel\Bundle\Bundle;
+
+abstract class Module extends Bundle
+{
+    abstract public function getShortName(): string;
+
+    public function build(ContainerBuilder $container)
+    {
+        $this->registerModuleControllers($container);
+    }
+
+    /**
+     * @param ContainerBuilder $container
+     * @return void
+     */
+    private function registerModuleControllers(ContainerBuilder $container): void
```

Comment:
We register controller services. The idea is to create similar methods for each service we want to define automatically (hooks/events, admin menus, etc.). The possibility to use a services.yml to load custom services will be available. But this methods exists to do the module development easiest.